### PR TITLE
Unity3.4.2

### PR DESCRIPTION
**Pull request by jindoujiao**
Update Unity Adapter to Unity SDK 3.4.2 with bug fixes and improvements.

### CODE REVIEW

File: adapters/Unity/UnityAdapter/GADMAdapterUnity.m [hunk 1]
```diff
@@ -81,17 +80,26 @@ - (void)getInterstitial {
...
   }
-  _isLoading = YES;
-  [[GADMAdapterUnitySingleton sharedInstance] requestInterstitialAdWithDelegate:self];
+
+  UnitySingletonCompletion completeBlock = ^(UnityAdsError *error, NSString *message) {
+    if(!error) {
+            [UnityAds addDelegate:self];
+            [UnityAds load:[self getPlacementID]];
+    } else {
+      id<GADMAdNetworkConnector> strongNetworkConnector = self->_networkConnector;
```
Comment (thatguyhampton) [inline adapters/Unity/UnityAdapter/GADMAdapterUnity.m]: Why use the `->` here and not `self._networkConnector`?

File: adapters/Unity/UnityAdapter/GADMAdapterUnity.m [hunk 2]
```diff
@@ -123,68 +131,74 @@ - (NSString *)getPlacementID {
...
   // notifies the Google Mobile Ads SDK accordingly.
-  if (strongNetworkConnector) {
+  if (strongNetworkConnector && [placementID isEqualToString:_placementID]) {
     [strongNetworkConnector adapterDidGetAdClick:self];
     [strongNetworkConnector adapterWillLeaveApplication:self];
   }
 }
 
 - (void)unityAdsDidError:(UnityAdsError)error withMessage:(NSString *)message {
+//  id<GADMAdNetworkConnector> strongNetworkConnector = _networkConnector;
```
Comment (thatguyhampton) [inline adapters/Unity/UnityAdapter/GADMAdapterUnity.m]: So I see the `NO_FILL` handled above. What happens when other errors occur? For instance what if we are already showing an AD?

File: adapters/Unity/UnityAdapter/GADMAdapterUnityRewardedAd.m [hunk 3]
```diff
@@ -81,27 +79,24 @@ - (void)requestRewardedAd {
...
-  }
-
-  [[GADMAdapterUnitySingleton sharedInstance] requestRewardedAdWithDelegate:weakSelf];
+  UnitySingletonCompletion completeBlock = ^(UnityAdsError *error, NSString *message) {
+        if(!error) {
+            [UnityAds addDelegate:self];
+            [UnityAds load:[self getPlacementID]];
+        } else {
+            NSError *errorWithDescription = GADUnityErrorWithDescription(message);
+            self->_adLoadCompletionHandler(nil, errorWithDescription);
```
Comment (thatguyhampton) [inline adapters/Unity/UnityAdapter/GADMAdapterUnityRewardedAd.m]: why the `->` and not dot syntax?

File: adapters/Unity/UnityAdapter/GADMAdapterUnityRewardedAd.m [hunk 4]
```diff
@@ -81,27 +79,24 @@ - (void)requestRewardedAd {
...
+            self->_adLoadCompletionHandler = nil;
```
Comment (thatguyhampton) [inline adapters/Unity/UnityAdapter/GADMAdapterUnityRewardedAd.m]: Same as above.

**Review by thatguyhampton — CHANGES_REQUESTED** [review]

File: adapters/Unity/UnityAdapter/GADMAdapterUnity.m (hunk 2 above)
Comment (serge-14) [reply adapters/Unity/UnityAdapter/GADMAdapterUnity.m]: `unityAdsDidFinish` with error status should be called.

File: adapters/Unity/UnityAdapter/GADMAdapterUnityRewardedAd.m (hunk 3 above)
Comment (serge-14) [reply adapters/Unity/UnityAdapter/GADMAdapterUnityRewardedAd.m]: c++ flashback :)

File: adapters/Unity/UnityAdapter/GADMAdapterUnitySingleton.m [hunk 5]
```diff
@@ -42,189 +33,60 @@ + (instancetype)sharedInstance {
...
   UADSMediationMetaData *mediationMetaData = [[UADSMediationMetaData alloc] init];
   [mediationMetaData setName:kGADMAdapterUnityMediationNetworkName];
   [mediationMetaData setVersion:kGADMAdapterUnityVersion];
   [mediationMetaData set:@"adapter_version" value:[UnityAds getVersion]];
   [mediationMetaData commit];
-
-  // Initializing Unity Ads with |gameID|.
-  [UnityAds initialize:gameID testMode:NO];
+  
+  [UnityAds initialize:gameID testMode:false enablePerPlacementLoad:true];
```
Comment (serge-14) [inline adapters/Unity/UnityAdapter/GADMAdapterUnitySingleton.m]: add delegate before?

File: adapters/Unity/UnityAdapter/GADMAdapterUnity.m (hunk 1 above)
Comment (serge-14) [reply adapters/Unity/UnityAdapter/GADMAdapterUnity.m]: I'm not a ObjC expert but is it correct way to convert weak to strong?

should it be something like:

if (self->_networkConnector) {
     __strong id<GADMAdNetworkConnector> strongNetworkConnector = self->_networkConnector;
}

**Pull request merged**